lazyflow/request/request.py

```diff
@@ -708,6 +708,21 @@ def _handle_finished_request(self, request, *args):
             if len(self.blocking_requests) == 0:
                 self._wake_up()
 
+    def add_done_callback(self, callback: Callable[["Request"], None]) -> None:
+        """
+        Mimic concurrent.Future interface
+        Call callback when request is finished regardless of the state
+        """
+        assert not self._cleaned, "This request has been cleaned() already."
+        with self._lock:
+            complete = self.execution_complete
```

Comment:
Control flow is not obvious here. It would be great to either add a comment or maybe use an early return:

```python
with self._lock:
    if not self.execution_complete:
        self._sig_execution_complete.subscribe(lambda: callback(self))
        return

callback(self)

Reply:
Updated :) not sure if added comment covers what you mean

Reply:
since it was copy-pasted from the one of the `notify_...` functions, it would make sense to have only a single pattern of control flow.

Reply:
OK, maybe it's not really necessary then. I was slightly confused by double-testing the `complete` variable within _and_ outside the lock, so I thought maybe it'd be better to have an early return.

Reply:
sure, I think this looks better, too. So we could change it in the whole file, no?

Reply:
> sure, I think this looks better, too. So we could change it in the whole file, no?

I think we should use the same pattern like in other notify functions, early return is not implementable for notifications where two flags are used https://github.com/ilastik/ilastik/pull/2390/files#diff-4801f2aba46cdd35eeca59573dbc9df361125acef1aad3e7a240928838a4f8f9R763

Reply:
okay, then let's leave it at that.